### PR TITLE
py/modsys.c: Add sys._exc_traceback.

### PR DESCRIPTION
**Pull request by DavidEGrayson**
This new function makes it easier to provide custom formatting of exception stack traces, for example to make them fit on a tiny 16x8-character display.  (Without this, I think you'd have to call `sys.print_exception`, somehow get the output as a string, and then scrape information from that string.)

I'm thinking of this as an experimental function that exposes internal details of MicroPython and therefore might change in the future.  That's why it has the underscore in the name.

Here is an example of its output:

    ['test.py', 4, 'foo', 'test.py', 7, '<module>']

The regular output from `sys.print_exception` for the same exception is:

```text
Traceback (most recent call last):
  File "test.py", line 7, in <module>
  File "test.py", line 4, in foo
ZeroDivisionError: divide by zero
```

Both outputs were generated by the following script:

```text
import sys

def foo():
  1/0

try:
  foo()
except Exception as e:
  sys.print_exception(e)
  print(sys._exc_traceback(e))
 ```

Is this the right approach for getting info about exception stack traces?   I'd be happy to change the name, change the output format, add tests, or add documentation.

### CODE REVIEW

**Comment by dpgeorge** [comment]
First thing to ask is how would one do this in CPython?

Then we need to consider if it's worth the cost (code size) and whether it should be guarded by a config macro.  I would say yes it should be, but then at what level would it be enabled, eg `MICROPY_CONFIG_ROM_LEVEL_EXTRA_FEATURES`.

**Comment by DavidEGrayson** [comment]
You can get this info in CPython is like this:

```py
import traceback

def foo():
    1/0

try:
    foo()
except Exception as e:
    tb = e.__traceback__
    while tb is not None:
        print("====")
        print("  filename: " + str(tb.tb_frame.f_code.co_filename))
        print("  line: " + str(tb.tb_lineno))
        print("  name: " + str(tb.tb_frame.f_code.co_name))
        tb = tb.tb_next
```

This involves at least three different classes: `traceback`, `frame`, and `code`.  The latter two contain quite a few things:

```text
dir(tb): ['tb_frame', 'tb_lasti', 'tb_lineno', 'tb_next']
dir(frame): ['__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'clear', 'f_back', 'f_builtins', 'f_code', 'f_globals', 'f_lasti', 'f_lineno', 'f_locals', 'f_trace', 'f_trace_lines', 'f_trace_opcodes']
dir(code): ['__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'co_argcount', 'co_cellvars', 'co_code', 'co_consts', 'co_filename', 'co_firstlineno', 'co_flags', 'co_freevars', 'co_kwonlyargcount', 'co_lines', 'co_linetable', 'co_lnotab', 'co_name', 'co_names', 'co_nlocals', 'co_posonlyargcount', 'co_stacksize', 'co_varnames', 'replace']
```

The Python standard library also comes with a [traceback module](https://github.com/python/cpython/blob/3.11/Lib/traceback.py) that introduces more classes.  MicroPython implements a [minimal traceback module](https://github.com/micropython/micropython-lib/blob/master/python-stdlib/traceback/traceback.py).

**Comment by DavidEGrayson** [comment]
I changed it so that this function is only enabled by default if `MICROPY_CONFIG_ROM_LEVEL_EXTRA_FEATURES` is 1.  (I'm glad to see this provides it for the RP2 boards.)  I also added documentation.

**Comment by dpgeorge** [comment]
Thanks for keeping this PR alive.  I can see how it would be useful.

To reduce code size I suggest implementing the functionality as an attribute of the exception type, rather than as a new function.  So it would go in `mp_obj_exception_attr()`.  To keep code size further down, it could reuse the existing `__traceback__` attribute name.  So the example from above would be:
```py
import sys

def foo():
  1/0

try:
  foo()
except Exception as e:
  sys.print_exception(e)
  print(e.__traceback__)
```

Now, that's not fully compatible with CPython, but it's a little bit closer than having a separate `sys._exc_traceback()` function.  We can say that if you want compatibility with CPython then you must consider `e.__traceback__` as an opaque (or implementation defined) value, and then use the `traceback` module to convert that opaque value to something usable.  Eg `traceback.extract_tb(e.__traceback__)` would work the same in CPython and MicroPython, even though the underlying value of `e.__traceback__` was different.

And then if you don't care about CPython compatibility and you want to efficiently get the traceback data, you can use the value of `e.__traceback__` directly (although maybe we should document it as opaque and subject to change).

File: py/objexcept.c [hunk 1]
```diff
@@ -266,9 +266,8 @@ mp_obj_t mp_obj_exception_get_value(mp_obj_t self_in) {
 
 void mp_obj_exception_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
     mp_obj_exception_t *self = MP_OBJ_TO_PTR(self_in);
-    if (dest[0] != MP_OBJ_NULL) {
```
Comment (dpgeorge) [inline py/objexcept.c:269]: I think this line needs to stay.  The code here must prevent store/delete from all attributes except `__traceback__`.

**Comment by DavidEGrayson** [comment]
I think this is ready to be merged in.  I changed it to make a list of tuples, so we can easily extend it in the future by adding to the tuple, and well-written user code will not break.  And there are tests and documentation now.

Perhaps the "minimal" build on Unix should not have this feature because it doesn't track line numbers.

File: tests/run-tests.py [hunk 2]
```diff
@@ -125,9 +128,7 @@
         "stress/list_sort.py",  # watchdog kicks in because it takes too long
     ),
     "minimal": (
-        "basics/class_inplace_op.py",  # all special methods not supported
-        "basics/subclass_native_init.py",  # native subclassing corner cases not support
-        "micropython/opt_level.py",  # don't assume line numbers are stored
+        "micropython/traceback.py",  # no line numbers, no list[N:] syntax
```
Comment (dpgeorge) [inline tests/run-tests.py:131]: Please don't remove these (see comment below).

File: tests/run-tests.py [hunk 3]
```diff
@@ -827,6 +828,8 @@ def run_tests(pyb, tests, args, result_dir, num_threads=1):
 
     # Skip platform-specific tests.
     skip_tests.update(platform_tests_to_skip.get(args.platform, ()))
+    if args.build == "minimal":
+        skip_tests.update(platform_tests_to_skip.get(args.build, ()))
```
Comment (dpgeorge) [inline tests/run-tests.py:832]: I understand why you made this change, but it's not the right approach.  We should not mix up the platform with the build.

If really needed we can add a `build_tests_to_skip` dict, to mirror `platform_tests_to_skip`, but for only one test I suggest following 7db50ccf5fd0932bacd40436c44d3283e1441d61 and adding the same code at the top of `traceback.py` to skip that test automatically.

File: tests/micropython/traceback.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# This comment allows test runners like run_script_on_remote_target
```
Comment (dpgeorge) [inline tests/micropython/traceback.py:1]: Suggest calling this test `exc_traceback.py`.

And then adding (or changing this comment) to briefly mention what it's testing, eg `Test values of BaseException.__traceback__` which are MicroPython specific".

**Comment by dpgeorge** [comment]
This increases the size of minimal ports by quite a bit.  But I think it's useful to have, and IMO more useful than `sys.print_exception()` which is enabled on all builds.  So I propose that we disable `sys.print_exception()` on minimal builds in favour of `exception.__traceback__`.  That should claw back a lot of the code size change.

But, let's do that in a separate PR, because it'll probably need changes to the test suite.

**Comment by dpgeorge** [comment]
@DavidEGrayson are you able to respond to the above review?  If not then I can make my suggested changes during rebase and merge of this PR.

**Comment by dpgeorge** [comment]
@DavidEGrayson I'll leave this up to you, let me know after you've had a chance to get back to it.

File: py/objexcept.c [hunk 5]
```diff
@@ -280,7 +280,23 @@ void mp_obj_exception_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
...
+        // custom formatting of the traceback by user code.
+        size_t entry_count = self->traceback_len / TRACEBACK_ENTRY_LEN;
+        size_t *data = self->traceback_data;
+        mp_obj_t obj = mp_obj_new_list(entry_count, NULL);
+        mp_obj_list_t *list = MP_OBJ_TO_PTR(obj);
+        for (size_t i = 0; i < entry_count; i++) {
+            size_t *src = &data[i * TRACEBACK_ENTRY_LEN];
+            mp_obj_t entry[3];
+            entry[0] = MP_OBJ_NEW_QSTR(src[0]);       // filename
+            entry[1] = MP_OBJ_NEW_SMALL_INT(src[1]);  // line number
```
Comment (dpgeorge) [inline py/objexcept.c:294]: If MICROPY_ENABLE_SOURCE_LINE is disabled then this number is always 1.  Maybe worth storing 0 (or -1?) instead, to indicate that line numbers are unavailable?

File: py/objexcept.c [hunk 6]
```diff
@@ -280,7 +280,23 @@ void mp_obj_exception_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
...
+            entry[2] = MP_OBJ_NEW_QSTR(src[2]);       // block
```
Comment (dpgeorge) [inline py/objexcept.c:295]: `src[2]` may be `MP_QSTRnull` if the block name is unknown, which shouldn't then be converted to a qstr object.

Not sure what the best way to handle this is.  Maybe change it so `MP_QSTR_` is used instead of `MP_QSTRnull` when the block name is unknown.

**Review by dpgeorge — COMMENTED** [review]
Regarding replacing `sys.print_exception(exc)` with this new `exc.__traceback__` feature: that might not work very well because `sys.print_exception` has a lot of nice features that aren't easily available with `exc.__traceback__`, such as errno name printing and nice formatting of the exception type and arguments.

I guess you could write Python to emulate that pretty printing.  Alternatively, could add another attribute to exceptions that return a pretty exception string, eg `exc.__str__` (to reuse a qstr).
